lib/portlayer/metrics/container.go

```diff
@@ -0,0 +1,327 @@
+func (cc *collectorVM) sampler() {
+	// TODO: replace this with service that only runs when there are
+	// subscribers
+	d := time.Duration(int64(sampleInterval)) * time.Second
+	for range time.Tick(d) {
+		// if we have no subscribers skip
+		if cc.SubscriberCount() == 0 {
+			continue
+		}
+		// collect metrics for current subscribers
+		cc.collect(cc.perfMgr, cc.Subscribers(), false)
+	}
+}
+
+// Sample returns a single metrics collection
+func (cc *collectorVM) Sample(contain interface{}) (chan interface{}, error) {
+	defer trace.End(trace.Begin(""))
+
+	containerSub, err := containerSubscriber(contain)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a manager for this request
+	mgr := performance.NewManager(cc.session.Vim25())
+
+	// create a publisher and subscribe
+	containerSub.publisher = pubsub.NewPublisher(100*time.Millisecond, 0)
+	ch := containerSub.publisher.Subscribe()
+
+	// create a map with this single subscriber
+	single := make(map[types.ManagedObjectReference]*subscriber)
+	single[containerSub.ref] = containerSub
+
+	// collect the metrics
+	go cc.collect(mgr, single, true)
+
+	return ch, nil
+}
+
+// collect will query vSphere for VM metrics and return to the subscribers
+func (cc *collectorVM) collect(m *performance.Manager, currentSubs map[types.ManagedObjectReference]*subscriber, single bool) {
+	defer trace.End(trace.Begin(""))
+
+	ctx := context.Background()
+	var mos []types.ManagedObjectReference
+
+	// metrics we are currently interested in monitoring
+	names := []string{"cpu.usagemhz.average", "mem.active.average"}
+
+	// create the spec
+	spec := types.PerfQuerySpec{
+		Format:     string(types.PerfFormatNormal),
+		MaxSample:  sampleSize,
+		IntervalId: sampleInterval,
+	}
+
+	// create a collection of object references
+	for mo := range currentSubs {
+		mos = append(mos, mo)
+	}
+
+	// if this is a single request close publisher when complete
+	if single {
+		defer currentSubs[mos[0]].publisher.Close()
+	}
+
+	// get the sample..
+	sample, err := cc.perfMgr.SampleByName(ctx, spec, names, mos)
+	if err != nil {
+		log.Errorf("unable to get metric sample: %s", err)
+		return
+	}
+	// convert to metrics
+	result, err := cc.perfMgr.ToMetricSeries(ctx, sample)
+	if err != nil {
+		log.Errorf("unable to convert metric sample to metric series: %s", err)
+		return
+	}
+
+	for i := range result {
+		met := result[i]
+		sub := currentSubs[met.Entity]
+		for s := range met.SampleInfo {
+			metric := VMMetrics{
+				CPU: CPUMetrics{
+					CPUs: []CPUUsage{},
+				},
+				Memory:     MemoryMetrics{},
+				SampleTime: met.SampleInfo[s].Timestamp,
+				Interval:   sampleInterval,
+			}
+			// the series will have values for each sample
+			for _, v := range met.Value {
+				switch v.Name {
+				case "cpu.usagemhz.average":
+					// convert to int so we skip the aggregate measurement
+					iid, err := strconv.Atoi(v.Instance)
+					if err != nil {
+						break
```

Comment:
I've updated this to check specifically for the aggregate instance and if not then convert.  If the conversion fails (highly doubtful) it will log the error and not publish the broken metric.

Additional comments added to provide background / clarity as well.